lib/miq_automation_engine/models/miq_ae_git.rb

```diff
@@ -0,0 +1,328 @@
+require 'rugged'
+
+class MiqAeGit
+  attr_accessor :name, :email, :base_name
+  ENTRY_KEYS = [:path, :dev, :ino, :mode, :gid, :uid, :ctime, :mtime]
+  DEFAULT_FILE_MODE = 0100644
+  LOCK_REFERENCE = 'refs/locks'
+
+  def initialize(options = {})
+    raise ArgumentError, "Must specify path" unless options.key?(:path)
+    @path          = options[:path]
+    @email         = options[:email]
+    @name          = options[:name]
+    @bare          = options[:bare]
+    @commit_sha    = options[:commit_sha]
+    @password      = options[:password]
+    @ssl_no_verify = options[:ssl_no_verify] || false
+    @remote_name   = 'origin'
+    @cred          = Rugged::Credentials::UserPassword.new(:username => @name,
+                                                           :password => @password)
+    @base_name     = File.basename(@path)
+    return clone(options[:url]) if options[:url]
+    options[:new] ? create_repo : open_repo
+  end
+
+  def create_repo
+    @repo = @bare ? Rugged::Repository.init_at(@path, :bare) :
+                    Rugged::Repository.init_at(@path)
+    @repo.config['user.name']       = @name  if @name
+    @repo.config['user.email']      = @email if @email
+  end
+
+  def open_repo
+    @repo = Rugged::Repository.new(@path)
+  end
+
+  def delete_repo
+    return false unless @repo
+    @repo.close
+    FileUtils.rm_rf(@path)
+    true
+  end
+
+  def add(hash, commit_sha = nil)
+    raise ArgumentError, "Must specify path" unless hash.key?(:path)
+    raise ArgumentError, "Must specify data" unless hash.key?(:data)
+    entry = {}
+    ENTRY_KEYS.each { |key| entry[key] = hash[key] if hash.key?(key) }
+    entry[:oid]  = @repo.write(hash[:data], :blob)
+    entry[:mode] ||= DEFAULT_FILE_MODE
+    entry[:mtime] ||= Time.now
+    current_index(commit_sha).add(entry)
+  end
+
+  def remove(path, commit_sha = nil)
+    current_index(commit_sha).remove(path)
+  end
+
+  def remove_dir(path, commit_sha = nil)
+    current_index(commit_sha).remove_dir(path)
+  end
+
+  def file_exists?(path, commit_sha = nil)
+    find_entry(path, commit_sha) ? true : false
+  end
+
+  def directory_exists?(path, commit_sha = nil)
+    entry = find_entry(path, commit_sha)
+    entry ? entry[:type] == :tree : false
+  end
+
+  def read_file(path, commit_sha = nil)
+    entry = find_entry(path, commit_sha)
+    raise MiqException::MiqGitEntryMissing, path unless entry
```

Comment:
This pattern of these two methods could move into a private method called `fetch_entry`

```ruby
def fetch_entry(path, commit_sha)
  find_entry(path, commit_sha).tap do |entry|
    raise MiqException::MiqGitEntryMissing, path unless entry
  end
end
```